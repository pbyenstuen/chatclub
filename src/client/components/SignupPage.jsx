import React, { useState } from "react";
import { useHistory } from "react-router";
import useSubmit from "./useSubmit";
import InputField from "./InputField";

const SignupPage = ({ api }) => {
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [confirm, setConfirm] = useState("");
    const [inputError, setInputError] = useState();
    const history = useHistory();

    const { handleSubmit: handleSignup, submitting, error } = useSubmit(
        async () => {
            await api.auth.signUp({
                username: username,
                password: password
            });
        },
        async () => {
            history.push("/login");
        }
    );

    const validateInput = (e) => {
        e.preventDefault();
        if (password !== confirm) { setInputError("Passwords do not match"); return; }
        if (!username || !password || !confirm) { setInputError("Please enter all fields"); return; }
        handleSignup(e);
    }

    return (
        <div id="signup-container">
            <header id="auth-header">
                <h2>SIGN UP</h2>
            </header>
            <form onSubmit={validateInput}>
                {submitting && <h4>Please wait...</h4>}
                {error && <h4>{error.toString()}</h4>}
                {inputError && <h4>{inputError.toString()}</h4>}
                <InputField
                    label={"Username"}
                    value={username}
                    onValueChange={setUsername}
                />
                <InputField
                    label={"Password"}
                    type={"password"}
                    value={password}
                    onValueChange={setPassword}
                />
                <InputField
                    label={"Confirm password"}
                    type={"password"}
                    value={confirm}
                    onValueChange={setConfirm}
                />
                <button className="submit-btn" type="submit" disabled={submitting}>SIGN UP</button>
            </form>
        </div>
    );
};

export default SignupPage;