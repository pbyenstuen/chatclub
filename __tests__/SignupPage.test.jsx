import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import SignupPage from "../src/client/components/SignupPage";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
}

describe("SignupPage", () => {
    it("can submit username and password", async () => {
        const signUp = jest.fn();

        const api = {
            auth: {
                signUp
            }
        }

        await mountComponent(<SignupPage api={api} />);

        Simulate.change(container.querySelector("input[name='username']"), {
            target: { value: "realDonaldTrump" },
        });

        Simulate.change(container.querySelector("input[name='password']"), {
            target: { value: "yourefired" },
        });

        Simulate.change(container.querySelector("input[name='confirmpassword']"), {
            target: { value: "yourefired" },
        });

        await act(async () => {
            Simulate.submit(container.querySelector("form"));
        });

        expect(signUp).toBeCalledWith({
            username: "realDonaldTrump",
            password: "yourefired"
        });
    });

    it("can show error message", async () => {
        const api = {
            auth: {
                signUp: () => {
                    throw new Error("Invalid username/password");
                }
            }
        }

        await mountComponent(<SignupPage api={api} />);

        container.querySelectorAll("input").forEach(e => 
            Simulate.change(e, {
            target: { value: "Value" },
        }));

        await act(async () => {
            Simulate.submit(container.querySelector("form"));
        });

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual("Error: Invalid username/password");
    });

    it("can show error message for empty fields", async () => {
        await mountComponent(<SignupPage />);

        Simulate.submit(container.querySelector("form"));

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual("Please enter all fields");
    });

    it("can show error message for password inequality", async () => {
        await mountComponent(<SignupPage />);

        Simulate.change(container.querySelector("input[name='username']"), {
            target: { value: "realDonaldTrump" },
        });

        Simulate.change(container.querySelector("input[name='password']"), {
            target: { value: "yourefired" },
        });

        Simulate.change(container.querySelector("input[name='confirmpassword']"), {
            target: { value: "yourfired" },
        });

        Simulate.submit(container.querySelector("form"));

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual("Passwords do not match");
    });
});