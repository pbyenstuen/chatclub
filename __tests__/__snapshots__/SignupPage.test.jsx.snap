// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupPage can show error message 1`] = `"<div id=\\"signup-container\\"><header id=\\"auth-header\\"><h2>SIGN UP</h2></header><form><h4>Error: Invalid username/password</h4><div><label>Username <input type=\\"text\\" name=\\"username\\" value=\\"Value\\"></label></div><div><label>Password <input type=\\"password\\" name=\\"password\\" value=\\"Value\\"></label></div><div><label>Confirm password <input type=\\"password\\" name=\\"confirmpassword\\" value=\\"Value\\"></label></div><button class=\\"submit-btn\\" type=\\"submit\\">SIGN UP</button></form></div>"`;

exports[`SignupPage can show error message for empty fields 1`] = `"<div id=\\"signup-container\\"><header id=\\"auth-header\\"><h2>SIGN UP</h2></header><form><h4>Please enter all fields</h4><div><label>Username <input type=\\"text\\" name=\\"username\\" value=\\"\\"></label></div><div><label>Password <input type=\\"password\\" name=\\"password\\" value=\\"\\"></label></div><div><label>Confirm password <input type=\\"password\\" name=\\"confirmpassword\\" value=\\"\\"></label></div><button class=\\"submit-btn\\" type=\\"submit\\">SIGN UP</button></form></div>"`;

exports[`SignupPage can show error message for password inequality 1`] = `"<div id=\\"signup-container\\"><header id=\\"auth-header\\"><h2>SIGN UP</h2></header><form><h4>Passwords do not match</h4><div><label>Username <input type=\\"text\\" name=\\"username\\" value=\\"realDonaldTrump\\"></label></div><div><label>Password <input type=\\"password\\" name=\\"password\\" value=\\"yourefired\\"></label></div><div><label>Confirm password <input type=\\"password\\" name=\\"confirmpassword\\" value=\\"yourfired\\"></label></div><button class=\\"submit-btn\\" type=\\"submit\\">SIGN UP</button></form></div>"`;
